### PR TITLE
[TRTLLM-8238][feat] Add EVS support for nano-v2-vlm

### PR DESCRIPTION
**Pull request by Wanli-Jiang**
## Summary by CodeRabbit

- New Features
  - Added optional video token pruning for multimodal inputs, controlled by an environment variable ratio (default off).
  - Multimodal outputs now include modality type and video size metadata for videos.
  - Introduced public utilities to compute retained token counts and retention masks for video embeddings.
- Behavior
  - When pruning is disabled, image and video processing remains unchanged.
- Tests/Chores/Docs
  - No user-facing changes reported.

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

### CODE REVIEW

File: tensorrt_llm/inputs/evs.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+        spatial_merge_size: The size of the spatial merge.
+        q: The pruning rate.
+
+    Returns:
+        The number of retained tokens.
+    """
+    T, H, W = map(int, video_size)
+    min_num_tokens = (H // spatial_merge_size) * (W // spatial_merge_size)
+    evs_num_tokens = int(T * min_num_tokens * (1 - q))
+    return max(min_num_tokens, evs_num_tokens)
```
[the remark this answers is not in the document]
Comment (BloodAxe) [reply tensorrt_llm/inputs/evs.py:32]: Just a side-note why `map(int,..)` exists here.
In VLLM we observed a rounding issue when input was Tensor versus when input was tuple of integers.
Tuple of ints input came from Preprocessing stage, while in actual forward() it was a Tensor. To make sure number of output tokens stays the case - an explicit cast was added.

File: tensorrt_llm/_torch/models/modeling_nanov2vlm.py [hunk 2]
```diff
@@ -304,32 +349,40 @@ def __call__(
...
+                t, _, h, w = processed_images['pixel_values'].shape
                 num_patches_list.append(processed_images['num_patches'])
                 pixel_values_list.append(processed_images['pixel_values'])
+                video_size_list.append([t, h, w])
 
                 # Processing the text prompt.
                 processed_query += parts[video_index]
-                for num_patches in processed_images['num_patches']:
+                total_num_patches = sum(processed_images['num_patches'])
+                desired_num_tokens = int(self.num_image_token *
```
Comment (BloodAxe) [inline tensorrt_llm/_torch/models/modeling_nanov2vlm.py:364]: Please don't use this way of computing output number of tokens. Instead use `compute_retained_tokens_count` from `evs.py`. It handles edge cases correctly (Always ensures we keep first frame without pruning)

File: tensorrt_llm/_torch/models/modeling_nanov2vlm.py [hunk 3]
```diff
@@ -117,6 +120,47 @@ def extract_feature(self, pixel_values):
...
+                for video_size in video_sizes:
+                    partial_mm_embed = mm_embed[start_idx:start_idx +
+                                                video_size[0]]
+                    retention_mask = compute_retention_mask(
+                        video_embeds=partial_mm_embed,
+                        video_size=video_size,
+                        spatial_merge_size=self.spatial_merge_size,
+                        q=VIDEO_PRUNING_RATIO,
+                        flatten_output=False,
+                    ).flatten(start_dim=1)
```
Comment (BloodAxe) [inline tensorrt_llm/_torch/models/modeling_nanov2vlm.py:155]: I'm not sure why you need to flatten here since `compute_retention_mask` operates on `[T, Hidden]` shape and returns mask of `[T]` shape. If you get different shape this probably indicate you're applying EVS to wrong input shape.

File: tensorrt_llm/_torch/models/modeling_nanov2vlm.py (hunk 2 above)
Comment (rakib-hasan) [reply tensorrt_llm/_torch/models/modeling_nanov2vlm.py:364]: When changing this, can we reuse the output from the calls inside `apply_evs()`?

File: tensorrt_llm/_torch/models/modeling_nanov2vlm.py [hunk 4]
```diff
@@ -304,32 +349,40 @@ def __call__(
...
-                    # NanoV2VL can only support PIL images. Convert normalized tensors (0-1) to PIL images (0-255).
-                    images = [
-                        Image.fromarray((image.permute(1, 2, 0) * 255).to(
-                            torch.uint8).cpu().numpy()) for image in video
-                    ]
-                else:
-                    images = video
                 # Processing for multimodal data.
-                processed_images = self.processor(images=images,
+                processed_images = self.processor(images=video,
```
Comment (rakib-hasan) [inline tensorrt_llm/_torch/models/modeling_nanov2vlm.py:353]: Is `video` here the frames or the whole videos? If frames, can we rename the tensor? If whole video, should we update the API from `images` to something like `media` ?

File: tensorrt_llm/inputs/evs.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+# and proprietary rights in and to this software, related documentation
+# and any modifications thereto.  Any use, reproduction, disclosure or
+# distribution of this software and related documentation without an express
+# license agreement from NVIDIA CORPORATION is strictly prohibited.
+
+import torch
+
+
+def compute_retained_tokens_count(video_size: torch.LongTensor,
+                                  spatial_merge_size: int, q: float) -> int:
```
Comment (rakib-hasan) [inline tensorrt_llm/inputs/evs.py:15]: is `q` commonly used for indicating pruning rate? Can we rename the argument to something more readable?

**Review by rakib-hasan — APPROVED** [review]
LGTM overall. Add some minor comments.

File: tensorrt_llm/_torch/models/modeling_nanov2vlm.py (hunk 4 above)
Comment (Wanli-Jiang) [reply tensorrt_llm/_torch/models/modeling_nanov2vlm.py]: it is list of images (splited from video input), and it is from HF codes definition, I will keep it unchanged since I cannot update it from trtllm side.

File: tensorrt_llm/_torch/models/modeling_nanov2vlm.py [hunk 6]
```diff
@@ -304,32 +419,42 @@ def __call__(
...
+                    # We need to know the length of the full input ids ahead, to make Mamba-mixer work.
+                    num_tokens_per_frame = [desired_num_tokens
+                                            ] + [0] * (num_frames - 1)
+                else:
+                    num_tokens_per_frame = [
+                        num_patches * self.num_image_token
+                        for num_patches in processed_images['num_patches']
+                    ]
+                for num_tokens in num_tokens_per_frame:
+                    image_repl = self.img_start_token + self.img_context_token * num_tokens + self.img_end_token
```
[the remark this answers is not in the document]
Comment (Wanli-Jiang) [reply tensorrt_llm/_torch/models/modeling_nanov2vlm.py]: added prompt `Frame {i}` between frames.

`fps` info is not available so no  "Frame {i} sampled at {t:.2f} sec" for now. Add a TODO comment for it.

**Comment by BloodAxe** [comment]
I do see presence of frame timestamps, and EVS here. However my knowledge of TRT-LLM internals is not enough to comment on changes outside of EVS. What I do expect to see is accuracy evaluations of Nano V2 on VideoMME & LongVideoBench[optional] with and without EVS. This would be the best smoke test of this MR.
What we are looking for is 128 frames max (2fps) setup with EVS 0, 0.50, 0.60, 0.70, 0.75, 0.80, 0.90.

File: tensorrt_llm/_torch/models/modeling_nanov2vlm.py [hunk 7]
```diff
@@ -107,17 +115,100 @@ def pixel_shuffle(self, x, scale_factor=0.5):
...
-        vit_embeds = self.pixel_shuffle(vit_embeds,
-                                        scale_factor=self.downsample_ratio)
-        vit_embeds = vit_embeds.reshape(vit_embeds.shape[0], -1,
-                                        vit_embeds.shape[-1])
-        vit_embeds = self.mlp1(vit_embeds)
+        # Use micro-batch to avoid OOM, especially for long video inputs.
+        micro_batch_size = 128
+        n = pixel_values.shape[0]
+        vit_embeds_lst = []
+        for i in range(0, n, micro_batch_size):
```
Comment (yechank-nvidia) [inline tensorrt_llm/_torch/models/modeling_nanov2vlm.py]: I think you need to address the last one, especially when n is not divisible by micro_batch_size.

File: tensorrt_llm/_torch/models/modeling_nanov2vlm.py [hunk 8]
```diff
@@ -273,64 +594,60 @@ def __call__(
...
+            ]
+            input_ids = torch.cat(input_ids_lst, dim=1)
+        else:
+            input_ids = self.tokenizer.encode(processed_query,
+                                              add_special_tokens=False,
+                                              return_tensors="pt")
 
         # Will package inputs for language model forward in AGGREGATE mode.
         multimodal_data = {}
         multimodal_data['pixel_values'] = processed_images['pixel_values'].to(
```
Comment (yechank-nvidia) [inline tensorrt_llm/_torch/models/modeling_nanov2vlm.py]: I missed here in previous PR, but can you do something like

multimodal_data['image']['pixel_values'] if image data is alive or
multimodal_data['video']['pixel_values'] if video_data is alive.

And other metadata, too. Refer to [https://github.com/NVIDIA/TensorRT-LLM/blob/main/tensorrt_llm/inputs/multimodal.py#L218-L226](https://github.com/NVIDIA/TensorRT-LLM/blob/main/tensorrt_llm/inputs/multimodal.py#L218-L226)

File: tensorrt_llm/_torch/models/modeling_nanov2vlm.py [hunk 9]
```diff
@@ -273,64 +594,60 @@ def __call__(
...
-                                          add_special_tokens=False,
-                                          return_tensors="pt")
+            processed_query, evs_ids = self._process_video_prompts(
+                split_text_prompt, num_tokens_per_frame_lst,
+                frame_separators_lst)
+
+        # Tokenize processed query.
+        if isinstance(processed_query, list):
+            input_ids_lst = [
+                self.tokenizer.encode(
```
Comment (yechank-nvidia) [inline tensorrt_llm/_torch/models/modeling_nanov2vlm.py]: I think I asked you this question before, but want to make sure that you are calling tokenizer.encode here because AutoProcessor.process() do not apply text encoding, right?

File: tensorrt_llm/_torch/models/modeling_nanov2vlm.py [hunk 10]
```diff
@@ -245,15 +351,230 @@ def _calculate_targets(
...
+            processed_query += image_repl + part
+
+        return processed_images, processed_query
+
+    def _process_videos_frames(
+            self, videos: List[Image.Image | torch.Tensor]) -> Dict[str, Any]:
+        num_patches_list = []
+        pixel_values_list = []
+        video_size_list = []
+        for video in videos:
```
[the remark this answers is not in the document]
Comment (yechank-nvidia) [reply tensorrt_llm/_torch/models/modeling_nanov2vlm.py]: Then, `for frame in video` would be more accurate in this sense.

File: tensorrt_llm/_torch/models/modeling_nanov2vlm.py [hunk 11]
```diff
@@ -245,15 +351,230 @@ def _calculate_targets(
...
+    def _split_video_content(
+        self, videos: Any
+    ) -> Tuple[List[List[Image.Image | torch.Tensor]], List[Dict[str, Any]
+                                                            | None]]:
+        video_frames = []
+        video_metadatas = []
+        for video_content in videos:
+            if isinstance(video_content, dict):
+                # Update it once video metadata is supported formally.
+                video_frames.append(video_content["frames"])
```
[the remark this answers is not in the document]
Comment (Wanli-Jiang) [reply tensorrt_llm/_torch/models/modeling_nanov2vlm.py]: yes, it is expected. Treated them as list of videos and process them one by one.

File: tensorrt_llm/_torch/models/modeling_nanov2vlm.py [hunk 12]
```diff
@@ -14,16 +14,22 @@
...
 from ..attention_backend import AttentionMetadata
 from ..model_config import ModelConfig
 from .modeling_auto import AutoModelForCausalLM
-from .modeling_multimodal_utils import find_input_mm_embeds, fuse_input_embeds
+from .modeling_multimodal_utils import (find_input_mm_embeds, fuse_input_embeds,
+                                        get_multimodal_embeddings)
 from .modeling_radio import RADIOVisionModel
 from .modeling_utils import register_auto_model
 
+VIDEO_PRUNING_RATIO = float(os.getenv("TLLM_VIDEO_PRUNING_RATIO", "0"))
```
[the remark this answers is not in the document]
Comment (Wanli-Jiang) [reply tensorrt_llm/_torch/models/modeling_nanov2vlm.py]: we will add to LLMArgs (with small knob if we have other group args)

File: tensorrt_llm/_torch/models/modeling_nanov2vlm.py (hunk 9 above)
Comment (Wanli-Jiang) [reply tensorrt_llm/_torch/models/modeling_nanov2vlm.py]: it is from convention-viewpoint, we can use from standard HF interface.

But for this model, we need to split it to support EVS.

**Review by kaiyux — APPROVED** [review]

File: tensorrt_llm/_torch/models/modeling_nanov2vlm.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (yechank-nvidia) [reply tensorrt_llm/_torch/models/modeling_nanov2vlm.py]: Then can you update the above type checking?

**Review by yechank-nvidia — APPROVED** [review]
